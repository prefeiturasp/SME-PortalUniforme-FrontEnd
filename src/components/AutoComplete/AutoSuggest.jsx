import API_FILACRECHE from "constants/apiFilaCreche.constants";
import React, { Component } from "react";
import Autosuggest from "react-autosuggest";

export class AutoSuggestAddress extends Component {
  constructor() {
    super();
    this.state = {
      value: "",
      addressObject: "",
      suggestions: []
    };
  }

  componentDidMount() {
    this.setState({
      value: this.props.value
    });
  }

  onChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  handleClearInput = () => {
    this.setState({
      value: '',
    });
  };

  componentDidUpdate(prevProps) {
    if (this.props.resetarFile && !prevProps.resetarFile) {
      this.handleClearInput();
    }
  }

  onSuggestionsFetchRequested = async ({ value }) => {
    if (value.length >= 4) {
      const response = await fetch(
        `${API_FILACRECHE.maps_api_endpoint}/q/${value.trim()}.js`
      );
      const json = await response.json();
      this.setState({ suggestions: json.results });
    }
  };

  getSuggestionValue = suggestion => {
    this.setState({ addressObject: suggestion });
    this.props.onAddressSelected(suggestion);
    return suggestion.street || suggestion.name || suggestion.display_name;
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  renderSuggestion(suggestion) {
    return <span>{suggestion.display_name}</span>;
  }

  render() {
    const { value, suggestions } = this.state;

    const inputProps = {
      placeholder: "Ex: Rua Doutor Diogo de Faria",
      value,
      type: "search",
      onChange: this.onChange,
      autoFocus: true,
      className: `form-control ${this.props.className}`,
      required: this.props.required,
      name: this.props.name
    };

    return (
      <Autosuggest
        suggestions={suggestions}
        onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.onSuggestionsClearRequested}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        inputProps={inputProps}
      />
    );
  }
}
