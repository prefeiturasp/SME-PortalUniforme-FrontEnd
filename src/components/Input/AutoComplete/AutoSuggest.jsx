import React, { Component } from "react";
import Autosuggest from "react-autosuggest";
import CorreiosService from "../../../services/Correios.service"

export class AutoSuggestAddress extends Component {
  constructor() {
    super();
    this.state = {
      value: "",
      addressObject: "",
      suggestions: []
    };
    this.correiosService = new CorreiosService()
  }

  componentDidMount() {
    this.setState({
      value: this.props.value
    });
  }

  onChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  handleClearInput = () => {
    this.setState({
      value: ""
    });
  };

  componentDidUpdate(prevProps) {
    if (this.props.resetarFile && !prevProps.resetarFile) {
      this.handleClearInput();
    }
  }

  fetchSuggestions = async (value, endpoint) => {
    let response = await fetch(
      `https://georef.sme.prefeitura.sp.gov.br/v1/${endpoint}?text=${value.trim()}&layers=address&boundary.gid=whosonfirst:locality:101965533`
    );
    let json = await response.json();
    return json.features;
  }

  fetchSuggestionsSearch = async (value, numberFound) => {
    const suggestions = await this.fetchSuggestions(value, 'search');
    const promises = suggestions.filter((s) => {
      let { housenumber, street, ...rest } = s.properties;
      if (!street) return;
      if (!housenumber && numberFound){
        housenumber = numberFound;
      }
      s.properties = { housenumber, street, ...rest };
      return s;
    })
    return await Promise.all(promises);
  }

  onSuggestionsFetchRequested = async ({ value }) => {
    if (value.length >= 4) {
      const execResp = /\d+/.exec(value);
      let suggestions;
      if (execResp) {
        const numeroBuscado = execResp[0];
        suggestions = await this.fetchSuggestionsSearch(value, numeroBuscado);
        if (suggestions.length === 0){
          suggestions = await this.fetchSuggestions(value, 'autocomplete');
        }
        if (suggestions.length !== 0){
          //Verifica se pelo menos uma das sugestões tem o número buscado
          let sugestaoBoa;
          suggestions = suggestions.filter((s) => {
            if (s.properties.housenumber === numeroBuscado){
              sugestaoBoa = s;
            } 
            else {
              return s;
            }
          })
          if (suggestions.length > 0 && !sugestaoBoa){
            const novaSugestao = suggestions[0];
            let { housenumber, label, name, ...rest } = novaSugestao.properties;
            housenumber = numeroBuscado;
            label = label.replace(/\d+/, numeroBuscado);
            name = name.replace(/\d+/, numeroBuscado);
            sugestaoBoa = {
              ...novaSugestao,
              properties: { housenumber, label, name, ...rest }
            };
          }
          suggestions.unshift(sugestaoBoa);
        }
      }
      else{
        suggestions = await this.fetchSuggestions(value, 'autocomplete');
        if (suggestions.length === 0){
          suggestions = await this.fetchSuggestionsSearch(value);
        }
      }
      const promises = suggestions.map(async (s) => {
        let { postalcode, neighbourhood, ...rest } = s.properties;
        const info = await this.correiosService.buscaInfo(
          s.properties.street, s.properties.housenumber);
        postalcode = info.cep
        neighbourhood = info.bairro
        s.properties = { postalcode, neighbourhood, ...rest };
        return s;
      })
      suggestions = await Promise.all(promises);
      this.setState({ suggestions });
    }
  };

  suggestionToString = (suggestion) => {
    const { street, housenumber, neighbourhood, postalcode, region, country } = suggestion.properties;
    let value = "";
    if (street){
      value += street;
    } 
    if (street && housenumber){
      value += ` ${housenumber}`;
    }
    value += ", ";
    if (neighbourhood) value += `${neighbourhood}, `;
    if (postalcode) value += `${postalcode}, `;
    if (region) value += `${region}, `;
    if (country) value += country;
    return value;
  }

  getSuggestionValue = suggestion => {
    this.setState({ addressObject: suggestion });
    this.props.onAddressSelected(suggestion);
    return this.suggestionToString(suggestion);
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  renderSuggestion = (suggestion) => {
    return (
      <span>{this.suggestionToString(suggestion)}</span>
    );
  }

  render() {
    const { value, suggestions } = this.state;
    const inputProps = {
      placeholder: "Ex: Rua Doutor Diogo de Faria",
      value,
      type: "search",
      onChange: this.onChange,
      className: `form-control ${this.props.className}`,
      required: this.props.required,
      name: this.props.name,
      onBlur: this.props.onAddressBlur
    };

    return (
      <Autosuggest
        suggestions={suggestions}
        onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.onSuggestionsClearRequested}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        inputProps={inputProps}
      />
    );
  }
}
