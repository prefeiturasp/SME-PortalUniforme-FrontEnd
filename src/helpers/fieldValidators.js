export const required = value =>
  value !== undefined ? undefined : "Campo obrigatório";

export const email = value =>
  value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value)
    ? "Email inválido"
    : undefined;

export const prefeituraEmail = value =>
  value && /.+@\prefeitura.sp.gov.br/.test(value)
    ? undefined
    : "Somente emails da prefeitura de São Paulo";

export const alphaNumeric = value =>
  value && /[^a-zA-Z0-9 ]/i.test(value)
    ? "Only alphanumeric characters"
    : undefined;

export const numericInteger = value =>
  value && /[^0-9 ]/i.test(value) ? "Somente números" : undefined;

export const phoneNumber = value =>
  value && !/^(0|[1-9][0-9]{9})$/i.test(value)
    ? "Invalid phone number, must be 10 digits"
    : undefined;
