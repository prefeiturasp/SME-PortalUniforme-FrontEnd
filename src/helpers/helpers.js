const STATUS_DICT = {
  INSCRITO: "Inscrição finalizada, em processo de credenciamento",
  CREDENCIADO: "Credenciado",
  EM_PROCESSO: "Em processo de inscrição, finalize o seu cadastro",
  PENDENTE: "Pendente credenciamento",
  ALTERADO: "Pendente validação de alteração",
};

export const getStatus = (status) => {
  return STATUS_DICT[status] || status;
};

export const getCNPJ = () => {
  return localStorage.getItem("cnpj");
};

export const getRazaoSocial = () => {
  return localStorage.getItem("razao_social");
};

export const getKey = (obj) => {
  return Object.keys(obj)[0];
};

export const getError = (obj) => {
  let result = "Erro";
  if (!obj[getKey(obj)]) {
    return "Erro";
  } else if (
    (obj[getKey(obj)][0] !== undefined &&
      typeof obj[getKey(obj)][0] !== "string") ||
    typeof obj[getKey(obj)] !== "string"
  ) {
    result = getError(obj[getKey(obj)]);
  } else {
    if (typeof obj[getKey(obj)] === "string") return obj[getKey(obj)];
    else return obj[getKey(obj)][0];
  }
  return result;
};
