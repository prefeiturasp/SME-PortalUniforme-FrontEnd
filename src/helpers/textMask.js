import { createTextMask } from "redux-form-input-masks";

export const fieldCPF_CNPJ = value => {
  if (value.length <= 11) {
    return value
      .replace(/\D/g, "")
      .replace(/(\d{3})(\d{3})(\d{3})(\d{2})/g, "$1.$2.$3-$4");
  } else {
    return value
      .replace(/\D/g, "")
      .replace(/(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})/g, "$1.$2.$3/$4-$5");
  }
};

export const fieldCep = createTextMask({
  pattern: "99999-999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});

export const fieldTel = createTextMask({
  pattern: "(99) 99999-9999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});

export const fieldCel = createTextMask({
  pattern: "(99) 99999-9999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});
