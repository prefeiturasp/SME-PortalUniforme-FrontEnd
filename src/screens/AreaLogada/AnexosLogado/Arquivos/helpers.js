export const verificarSeFaltamArquivos = (empresa, tiposDocumentos) => {
  let aindaFaltaDocumentoObrigatorio = false;
  tiposDocumentos.forEach((tipoDocumento) => {
    if (
      tipoDocumento.obrigatorio &&
      !empresa.arquivos_anexos.find(
        (arquivo) => arquivo.tipo_documento.id === tipoDocumento.id
      )
    )
      aindaFaltaDocumentoObrigatorio = true;
  });
  let aindaFaltamArquivos =
    empresa.lojas.find((loja) => loja.foto_fachada === null) ||
    empresa.arquivos_anexos.length === 0 ||
    aindaFaltaDocumentoObrigatorio;
  return aindaFaltamArquivos;
};
