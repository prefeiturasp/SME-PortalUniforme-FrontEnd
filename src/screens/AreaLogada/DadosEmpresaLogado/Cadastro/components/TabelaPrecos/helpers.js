export const getPrecoVezesQuantidade = (values, uniforme) => {
  if (values[uniforme.nome]) {
    const valor =
      parseFloat(values[uniforme.nome].replace(",", ".")) * uniforme.quantidade;
    return valor
      .toFixed(2)
      .toString()
      .replace(".", ",");
  } else {
    return "0,00";
  }
};

export const getTotal = (tipoDeUniforme, values) => {
  let total = 0;
  tipoDeUniforme.uniformes.forEach((uniforme) => {
    if (values[uniforme.nome]) {
      total +=
        parseFloat(values[uniforme.nome].replace(",", ".")) *
        uniforme.quantidade;
    }
  });
  return total
    .toFixed(2)
    .toString()
    .replace(".", ",");
};

export const maiorQueLimite = (tipoDeUniforme, values, limites) => {
  let total = 0;
  tipoDeUniforme.uniformes.forEach((uniforme) => {
    if (values[uniforme.nome]) {
      total +=
        parseFloat(values[uniforme.nome].replace(",", ".")) *
        uniforme.quantidade;
    }
  });
  const preco_maximo = limites.find(
    (limite) => limite.categoria_uniforme === tipoDeUniforme.id
  ).preco_maximo;
  if (total > parseFloat(preco_maximo)) {
    return preco_maximo.replace(".", ",");
  } else return false;
};
