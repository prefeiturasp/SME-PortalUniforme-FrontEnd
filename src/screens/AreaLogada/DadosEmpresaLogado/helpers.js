import { maiorQueLimite } from "./Cadastro/components/TabelaPrecos/helpers";

const addCidadeEstadoSP = (empresa) => {
  empresa.lojas.map((loja) => {
    loja.cidade = "São Paulo";
    loja.uf = "SP";
    return loja;
  });
  return empresa;
};

const formataPrecos = (empresa) => {
  let categorias = [];
  empresa.ofertas_de_uniformes.forEach((oferta) => {
    if (!categorias.includes(oferta.uniforme_categoria)) {
      categorias.push(oferta.uniforme_categoria);
    }
  });
  categorias.forEach((categoria) => {
    empresa[categoria] = true;
  });
  empresa.ofertas_de_uniformes.forEach((oferta) => {
    empresa[oferta.nome] = oferta.preco.replace(".", ",");
  });
  return empresa;
};

export const formataEmpresa = (empresa) => {
  empresa = addCidadeEstadoSP(empresa);
  empresa = formataPrecos(empresa);
  return empresa;
};

export const formataPayloadLojasPrecos = (values, tiposDeUniforme) => {
  let ofertas_de_uniformes = [];
  tiposDeUniforme.forEach((tipoDeUniforme) => {
    tipoDeUniforme.uniformes.forEach((uniforme) => {
      if (values[uniforme.nome]) {
        ofertas_de_uniformes.push({
          nome: uniforme.nome,
          valor: parseFloat(values[uniforme.nome].replace(",", ".")),
        });
      }
    });
  });
  values.ofertas_de_uniformes = ofertas_de_uniformes;
  return values;
};

export const validaTabelaPrecos = (values, tiposDeUniforme, limites) => {
  let erro = false;
  if (
    tiposDeUniforme.filter((tipoDeUniforme) => values[tipoDeUniforme.id])
      .length === 0
  ) {
    erro = "É necessário fornecer ao menos um grupo de uniformes escolares";
    return erro;
  }
  tiposDeUniforme.filter(tipo => tipo.uniformes.length > 0).forEach((tipoDeUniforme) => {
    if (
      limites.find((value) => value.categoria_uniforme === tipoDeUniforme.id)
        .obrigatorio &&
      maiorQueLimite(tipoDeUniforme, values, limites)
    ) {
      erro = `O valor máximo para os itens em ${
        tipoDeUniforme.nome
      } é R$ ${maiorQueLimite(tipoDeUniforme, values, limites)}`;
      return erro;
    }
  });
  return erro;
};
