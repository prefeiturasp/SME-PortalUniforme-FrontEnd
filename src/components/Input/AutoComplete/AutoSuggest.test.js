import { mount } from "enzyme";
import React from "react";
import { required } from "helpers/fieldValidators";
import { AutoSuggestAddress } from "./AutoSuggest";

describe("test <AutoSuggestAddress>", () => {
  let wrapper;
  let onChange = jest.fn();
  let input = { onChange };

  beforeEach(() => {
    const props = {
      meta: {
        touched: true,
        error: "This field is required"
      }
    };
    wrapper = mount(
      <AutoSuggestAddress
        className="teste"
        name="myCombo"
        input={input}
        label="mylabel"
        required
        value=""
        validate={required}
        {...props}
      />
    );
  });

  it("renders component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
