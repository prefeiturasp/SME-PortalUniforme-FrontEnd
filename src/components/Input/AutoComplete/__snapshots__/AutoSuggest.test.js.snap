// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <AutoSuggestAddress> renders component 1`] = `
<AutoSuggestAddress
  className="teste"
  input={
    Object {
      "onChange": [MockFunction],
    }
  }
  label="mylabel"
  meta={
    Object {
      "error": "This field is required",
      "touched": true,
    }
  }
  name="myCombo"
  required={true}
  validate={[Function]}
  value=""
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "autoFocus": true,
        "className": "form-control teste",
        "name": "myCombo",
        "onChange": [Function],
        "placeholder": "Ex: Rua Doutor Diogo de Faria",
        "required": true,
        "type": "search",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  >
    <Autowhatever
      containerProps={Object {}}
      getSectionItems={[Function]}
      highlightedItemIndex={null}
      highlightedSectionIndex={null}
      id="1"
      inputProps={
        Object {
          "autoFocus": true,
          "className": "form-control teste",
          "name": "myCombo",
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyDown": [Function],
          "placeholder": "Ex: Rua Doutor Diogo de Faria",
          "required": true,
          "type": "search",
          "value": "",
        }
      }
      itemProps={[Function]}
      items={Array []}
      multiSection={false}
      renderInputComponent={[Function]}
      renderItem={[Function]}
      renderItemData={
        Object {
          "query": "",
        }
      }
      renderItemsContainer={[Function]}
      renderSectionTitle={[Function]}
      theme={
        Object {
          "container": "react-autosuggest__container",
          "containerOpen": "react-autosuggest__container--open",
          "input": "react-autosuggest__input",
          "inputFocused": "react-autosuggest__input--focused",
          "inputOpen": "react-autosuggest__input--open",
          "item": "react-autosuggest__suggestion",
          "itemFirst": "react-autosuggest__suggestion--first",
          "itemHighlighted": "react-autosuggest__suggestion--highlighted",
          "itemsContainer": "react-autosuggest__suggestions-container",
          "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
          "itemsList": "react-autosuggest__suggestions-list",
          "sectionContainer": "react-autosuggest__section-container",
          "sectionContainerFirst": "react-autosuggest__section-container--first",
          "sectionTitle": "react-autosuggest__section-title",
        }
      }
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-owns="react-autowhatever-1"
        className="react-autosuggest__container"
        key="react-autowhatever-1-container"
        role="combobox"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls="react-autowhatever-1"
          autoComplete="off"
          autoFocus={true}
          className="form-control teste"
          key="react-autowhatever-1-input"
          name="myCombo"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Ex: Rua Doutor Diogo de Faria"
          required={true}
          type="search"
          value=""
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1"
          key="react-autowhatever-1-items-container"
          role="listbox"
        />
      </div>
    </Autowhatever>
  </Autosuggest>
</AutoSuggestAddress>
`;
